src/rgw/rgw_aws_sns.cc

```diff
@@ -0,0 +1,199 @@
+#include <string>
+#include <iostream>
+#include <utility>
+#include "common/dout.h"
+#include <aws/core/Aws.h>
+#include <aws/sns/SNSClient.h>
+#include <boost/lockfree/queue.hpp>
+#include <aws/core/utils/ARN.h>
+#include <aws/core/auth/AWSCredentials.h>
+#include <unordered_set>
+#include <aws/sns/model/PublishRequest.h>
+
+namespace rgw::sns {
+
+  static const int STATUS_OK = 1;
+  static const int STATUS_FULL = 2;
+  static const int STATUS_MANAGER_STOPPED = 3;
+  static const int STATUS_CLIENT_CLOSED = 4;
+
+  struct client_id_t {
+    const std::string &accessKey;
+    const std::string &accessSecret;
+    const std::string &arn;
+    bool useSSL;
+
+    client_id_t(const std::string &accessKey, const std::string &accessSecret,
+                const std::string &arn, bool useSsl) : accessKey(accessKey),
+                                                       accessSecret(accessSecret),
+                                                       arn(arn), useSSL(useSsl) {}
+
+    bool operator==(const client_id_t &other) const {
+      return accessKey == other.accessKey &&
+             accessSecret == other.accessSecret &&
+             arn == other.arn && useSSL == other.useSSL;
+    }
+
+    struct hasher {
+      std::size_t operator()(const client_id_t &k) const {
+        return ((std::hash<std::string>()(k.accessKey)
+                 ^ (std::hash<std::string>()(k.accessSecret) << 1)) >> 1)
+               ^ (std::hash<std::string>()(k.arn) << 1)
+               ^ (std::hash<bool>()(k.useSSL) << 1);
+      }
+    };
+
+
+  };
+
+  struct message_t {
+    const Aws::SNS::SNSClient *client;
+    const std::string message;
+    const std::string topic;
+
+    message_t(const Aws::SNS::SNSClient *client, std::string message, std::string topic) :
```

Comment:
another option is to have:
```
message_t(const Aws::SNS::SNSClient *client, const std::string& message, const std::string& topic) :
  client(client),
￼  message(message),
￼  topic(topic) {};
``` 
in both cases there is only one string copy. do you think there is extra value in using the one with `std::move()`?